<a href="https://colab.research.google.com/github/MussaddikKhan/Python-College-Experiments/blob/master/FYMCA/updated/Experiment_No_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Expeirment No 7:**
***`Create a Python class representing a shape and derive subclasses like rectangle, circle, and triangle. Each subclass should have methods to calculate its area and perimeter.`***




In [7]:
# Base class: Shape
class Shape:
    def area(self):
        pass

    def perimeter(self):
        pass

# Rectangle class derived from Shape
class Rectangle(Shape):
    def __init__(self, length, width):
        self.length = length
        self.width = width

    def area(self):
        return self.length * self.width

    def perimeter(self):
        return 2 * (self.length + self.width)

# Circle class derived from Shape
import math
class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius

    def area(self):
        return math.pi * self.radius**2

    def perimeter(self):
        return 2 * math.pi * self.radius

# Triangle class derived from Shape
class Triangle(Shape):
    def __init__(self, a, b, c):
        self.a = a
        self.b = b
        self.c = c

    def area(self):
        # Using Heron's formula for area
        s = (self.a + self.b + self.c) / 2
        return math.sqrt(s * (s - self.a) * (s - self.b) * (s - self.c))

    def perimeter(self):
        return self.a + self.b + self.c

# Create objects for each shape
rect = Rectangle(5, 3)
circle = Circle(7)
triangle = Triangle(3, 4, 5)

# Display the area and perimeter of each shape
print(f"Rectangle: Area = {rect.area()}, Perimeter = {rect.perimeter()}")
print(f"Circle: Area = {circle.area()}, Perimeter = {circle.perimeter()}")
print(f"Triangle: Area = {triangle.area()}, Perimeter = {triangle.perimeter()}")


Rectangle: Area = 15, Perimeter = 16
Circle: Area = 153.93804002589985, Perimeter = 43.982297150257104
Triangle: Area = 6.0, Perimeter = 12


### Explanation:

######Base Class (`Shape`):
- **`class Shape:`**  
  This is the base class that defines the methods `area()` and `perimeter()`. These methods are not implemented in the base class, but are expected to be overridden by subclasses.

###### Rectangle Class:
- **`class Rectangle(Shape):`**  
  The `Rectangle` class inherits from `Shape`. It has a constructor that accepts the `length` and `width` of the rectangle.  
  - **`area(self)`**: This method calculates the area of the rectangle using the formula `length * width`.
  - **`perimeter(self)`**: This method calculates the perimeter of the rectangle using the formula `2 * (length + width)`.

###### Circle Class:
- **`class Circle(Shape):`**  
  The `Circle` class inherits from `Shape`. It takes the `radius` as input.  
  - **`area(self)`**: The area of a circle is calculated using the formula `π * radius^2` (using `math.pi` for the value of π).
  - **`perimeter(self)`**: The perimeter (circumference) of the circle is calculated using the formula `2 * π * radius`.

###### Triangle Class:
- **`class Triangle(Shape):`**  
  The `Triangle` class inherits from `Shape` and takes the lengths of the three sides (`a`, `b`, `c`) as input.  
  - **`area(self)`**: The area of the triangle is calculated using **Heron’s formula**. First, the semi-perimeter `s` is calculated as `(a + b + c) / 2`. Then, the area is calculated using the formula:  
    \[
    \text{Area} = \sqrt{s(s - a)(s - b)(s - c)}
    \]
  - **`perimeter(self)`**: The perimeter of the triangle is the sum of the three sides: `a + b + c`.

###### Example:
1. **Rectangle** with `length = 5` and `width = 3`
   - Area = `5 * 3 = 15`
   - Perimeter = `2 * (5 + 3) = 16`

2. **Circle** with `radius = 7`
   - Area = `π * 7^2 = 153.938`
   - Perimeter = `2 * π * 7 = 43.982`

3. **Triangle** with sides `a = 3`, `b = 4`, and `c = 5`
   - Area = `6` (using Heron's formula)
   - Perimeter = `3 + 4 + 5 = 12`
